### ISSUE TITLE
[New build - EXPERIMENTAL] 21w28a: Docking Port and sub-modules

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#3991]
This build introduce the new docking port and a new Docking Controller allowing you to spawn decorative prefabs in a random fashion.
Check out the upcoming post to learn more about the new docking philosophy.

Docking New Features
- New Docking Connector brick. It can grid-resize.
- New Docking Controller brick.
- Configure "spawning sub-modules" inside the controller.
- Add tags to be spawned with a given weight for the randomized draw.
- Any entity with a docking module can be used as a sub-module, to be configured inside the codex entity windows.

Still to do: 
- Connector size check for spawn
- Station Module docking uses new docking port.
- Kiosks spawning for stations
- Spaceship universal docking port.

Also new due to popular request:
- Chain link: simply holding Left-SHIFT during a link action will hold-on to the current link.
- Box delete tool: select a volume like with the copy/paste. Then right click to delete the selection.
- F3 Option: disable NPC spawn
- F3 Option: peaceful NPCs mode.

Bugfixes:
#3987 Station Starting block name.

**Comment by ZachZent** [comment]
I love it! Since you are doing connectors and whatnot, I'd though I'd repost an idea about a year ago that you quite liked. https://github.com/tsunamayo/Starship-EVO/issues/2817

> To add, I'd like to see a 30, 45, 60 degree connector rather than only straight connectors. As great as they are, you are kind of limited to square-like stations and any angle requires an annoyingly long time to angle even a basic hallway if it is possible in the first place. That is unless you want to build an entire module on a child entity, but then the union will come out and break your legs. Since a module is a new grid anyway, why not skip the child entity middle man allowing for easier creation of more "rounded" stations. For example something like this where as before you would only be able to do 4 comfortably.
![image](https://user-images.githubusercontent.com/36980446/125409180-7ee96780-e370-11eb-9fbd-c8c33b1f43e4.png)

To add to it, during the station comp we found that all the stations when "generated" had a tendency to be very grid like with large spaces in between them all. You can see it clearly with Dwarf's station already in game. The only exception was a station built by Domovoi where the modules were so close together that it looked solid. It will help vary up the stations as well as give more options in design

**Comment by Dwarf-LordPangolin** [comment]
Oh wow! Tons of cool stuff here!

I'm fine with the blueprint being shared; I just ask people to bear in mind that it's very much a work in progress, and the current version isn't finished. 😄

**Comment by tsunamayo** [comment]
@Dwarf-LordPangolin Thanks mate!
Here we go:
[Sub-modules.zip](https://github.com/tsunamayo/Starship-EVO/files/6808486/Sub-modules.zip)
That must be a bug, let me check.

@ZachZent yes I agree it is very square-y. Not too deal breaker from my side, but I agree if we could try to make something like you suggested also for further design that would be very cool. 

Why using a child entity bother you? Child entity impact has been massively reduced now, it is not 0 but pretty close to it (I still want to fight decorative child creep though). So no impact from a performance side should you guys want to try this.
Maybe it is to have a perfectly aligned docked entity? 

This is something I want to tackle at one point anyway. As you know I want us to get to the point where the torus would be made of brick entities. I dont want a single piece as that would not work, so I wanted to spawn many sections and dock them at a slight angle to make the ring. Maybe a wedged connector with an angle and an offset parameter?
But there would be many issues due to seams on the hull... That could be a case for allowing clipping. Ah clipping, our old friend...
Or what about an accordion type of corridor or brick that self adjust between the entity? On paper that would use the same tech as my yet to be redone soft piping (spline interpolation inside the vertex shader)

Also one thing we really need is some continuous circular motion. As the collisions has been dramatically improved I feel quite confident to put that in the game now...

**Comment by timethatswasted** [comment]
> But there would be many issues due to seams on the hull... That could be a case for allowing clipping. Ah clipping, our old friend...


Just let it happen. You will make a lot of people happy.  

I would think at this point it’s more memory intensive to calculate collisions to stop clipping and have people adding layers of slabs, 1/8 blocks and child entities to cover up the seam…. then it is to just allow clipping and find a work around to any exploits.

also it’s just sexy.

**Comment by ZachZent** [comment]
> Why using a child entity bother you? Child entity impact has been massively reduced now, it is not 0 but pretty close to it (I still want to fight decorative child creep though). So no impact from a performance side should you guys want to try this.

Most of it is old thinking really and there is nothing stopping me from using CEs. Part of it though is from experience trying to make shapes using hinges or rotators. I've found that without extensive preplanning, using CEs as is rather imprecise. A hinge or rotator at the vertices of the shape will often leave an odd gap impossible to completely fill with bricks. Even with the 1/8 we have now, the only way I was able to do that on the fly is have two sliders under the hinge or rotator to put it into the right place which required a lot of trial and error. You can have a rotator at the very center of a shape which is more reliable and is easier to preplan, but then you have a number of rotators in the center and have to do something about them. The connector idea I had could be 100% configurable like a hinge rather then a preset 30, 45, and 60 or something like the accordion which sounds a bit like the old hose tech (which totally should make a come back). At this point, those individually would be the easiest to work with

**Comment by tsunamayo** [comment]
@timethatswasted no I dont think it is sexy, I think clipping look like crap. Especially when people dont care and leave z-figthing  surfaces everywhere. I dont get why people try so hard to make smooth surface when it never look as good as a more angular one.
And there is some gameplay consideration behind, so sorry that train left...

@ZachZent I dont like the current torus because:
- thats some specific unoptimized code rendering and collision wise. I want to get rid of that code.
- It has a massive cost on my side if I want to add some variety. I dont think it is cool to have player explore the galaxy to always stumble on the same ring.
- For the smooth side, I actually count it as a neg. It doesnt match the style of the game. Plus a voxel prefab would still look good if you dock 30+ sections
- We could do some much better with bricks, with some nice details, and also add a lot of variety...
- Air shader is laggy and too complex, I will likely get rid of it. It wouldnt look as good, but would fit all design. Current is just impossible to make evolve.
- Yes I need to redo the tech for ground, using voxels (a different kind though, with LOD). I am still not sure if each section would have its own voxel world or if I would just spawn on single for all the hull sections. In anycase they would all use the same procedural terrain so they would match.

**Comment by Kaiser-Indrasil** [comment]
> I dont get why people try so hard to make smooth surface when it never look as good as a more angular one.

@tsunamayo I beg to differ. Just look at how godlike Ulty's stuff is:

![image](https://user-images.githubusercontent.com/63756057/125534228-06dc0b38-68bd-41e6-ad40-9435d7345805.png)

**Comment by ZachZent** [comment]
Fair enough. A benefit with a completely custom ring would be an internal structure inside the ring a la forerunners from halo. Either in this post or a separate discussion, why don't you share your idea for terrain as it seems to have deviated from the expected.

Though to shift back to the original purpose, I'm not sure if variable degree connectors would be the best option in general. For the future ring it is the best option rather then using hinges. Much less work on the builder's part and allows variability like station generations. However there is the issue with filling in the gaps between connected areas. Only some angles would be able to be completely filled in without clipping or gaps and besides the 30, 45, 60, the effort would be quite large if you need to fill in the gaps with 1/8 blocks on 21km ring would be extreme.
![image](https://user-images.githubusercontent.com/36980446/125536484-ff9e7054-5dab-465e-82bd-47601ea34a77.png)
Quick and dirty image of a 30 degree connector. With this it is easy to fill in the gaps since it is just 30 degrees, the blocks stretches out to fill it in well both on the open and closed side. But imagine an 18 degree connector which would make a 20 sided polygon. Much tougher if you want to avoid gaps and clipping. I'm not sure of the answer truthfully as hinges/rotators have the same problem, but a way to allow the connection of angled entities will greatly enhance the stations and the future ring.

**Comment by Garrett-C** [comment]
> Fair enough. A benefit with a completely custom ring would be an internal structure inside the ring a la forerunners from halo. Either in this post or a separate discussion, why don't you share your idea for terrain as it seems to have deviated from the expected.
> 
> Though to shift back to the original purpose, I'm not sure if variable degree connectors would be the best option in general. For the future ring it is the best option rather then using hinges. Much less work on the builder's part and allows variability like station generations. However there is the issue with filling in the gaps between connected areas. Only some angles would be able to be completely filled in without clipping or gaps and besides the 30, 45, 60, the effort would be quite large if you need to fill in the gaps with 1/8 blocks on 21km ring would be extreme.
> ![image](https://user-images.githubusercontent.com/36980446/125536919-1f8a00e2-4c31-4f60-9178-b3a01c4284dc.png)
> Quick and dirty image of a 30 degree connector. With this it is easy to fill in the gaps since it is just 30 degrees, the blocks stretches out to fill it in well both on the open and closed side. But imagine an 18 degree connector which would make a 20 sided polygon. Much tougher if you want to avoid gaps and clipping. I'm not sure of the answer truthfully as hinges/rotators have the same problem, but a way to allow the connection of angled entities will greatly enhance the stations and the future ring.

Not sure what you mean about angles other than those 3 you listed being difficult. Any ratio of whole numbers between 1:1 and 1:16 would be easily doable with out clipping. So an adjustable angle on a docking tube would make sense to accommodate that.

**Comment by ZachZent** [comment]
> Not sure what you mean about angles other than those 3 you listed being difficult. Any ratio of whole numbers between 1:1 and 1:16 would be easily doable with out clipping. So an adjustable angle on a docking tube would make sense to accommodate that.

Good point, I was more thinking if you were to choose the exact angle like you would on a hinge.
![image](https://user-images.githubusercontent.com/36980446/125544491-a6c072a1-7c21-4415-9558-0244df428bc0.png)
For example this 100 sided polygon would require center rotator increments of 3.6 degrees of exterior hinge/connector angles of 176.4 degrees. Filling in the gaps on the outside and slimming in the connected parts on the inside if the build goes further in from the connected point may not be possible using only stretched blocks without leaving a gap or clipping (though I have not done the math to 100% prove)

I'd expect the custom torus to have many more sides as to at least keep the general appearance of smoothness. even if not a perfect circle. Maybe if connector angle increments you can select are limited to what can be filled in using stretched blocks to the max stretch of 1x16 displayed with the number of sides you will get, that will be functional enough for players to make these many sided polygons knowing that the gaps can be filled in.
![image](https://user-images.githubusercontent.com/36980446/125546248-60fb2e9f-c723-41b1-b02c-6af13d2136a7.png)

**Comment by Uncle-Ulty** [comment]
> I dont get why people try so hard to make smooth surface when it never look as good as a more angular one.

Ouch! direct critical hit!


> I beg to differ. Just look at how godlike Ulty's stuff is: thanks <3. (Every time I see this ship I get annoyed about some minor details that should be improved)


about torus stations, I don't see any angle that allows smooth surfaces. maybe a combination of 2 or more angles would work, but It would make everything harder to make... I think the cosmetic solution is an outer or inner plate to fill de gaps...

a 1:16 wedge would work for Zach sample of a 100 sided polygon, since the 1:16 angle is 3,57°. It still would remain a small gap of 0,03º per section, but it would be visually negligible.

All the possible wedge angles of bricks proportions allowed in this game:

![image](https://user-images.githubusercontent.com/73253472/125651369-272a4bb3-73a4-44ca-bb78-7c838661af2b.png)

**Comment by ZachZent** [comment]
For future reference, I expanded on Ulty's graph including the more specific angles and the accompanying external angle.
![image](https://user-images.githubusercontent.com/36980446/125897225-b3bfd280-cd77-4430-b1ca-feecc258d902.png)

**Comment by tsunamayo** [comment]
@Uncle-Ulty @ZachZent  @Garrett-C Haha indeed you guys are amazing. Yes we could make those neat angle, I wonder if it is too complex or not at the end. Indeed what I would do is a skirt to cover the seam. I will implement a recursive spawn algo so we can start toying with that.

**Comment by Garrett-C** [comment]
Sounds like you have some neat ideas for them. Very excited.

**Comment by timethatswasted** [comment]
Hmm can we also get a “skirt” to cover the seems of child entities for ships and stations?  I’ve been using pipes since they clip through blocks to fill the gaps. A square or wedge shaped pipe would be nice….
